lib/public_key/src/pubkey_ocsp.erl

```diff
@@ -0,0 +1,290 @@
+%%
+%% %CopyrightBegin%
+%%
+%% Copyright Ericsson AB 2008-2020. All Rights Reserved.
+%%
+%% Licensed under the Apache License, Version 2.0 (the "License");
+%% you may not use this file except in compliance with the License.
+%% You may obtain a copy of the License at
+%%
+%%     http://www.apache.org/licenses/LICENSE-2.0
+%%
+%% Unless required by applicable law or agreed to in writing, software
+%% distributed under the License is distributed on an "AS IS" BASIS,
+%% WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+%% See the License for the specific language governing permissions and
+%% limitations under the License.
+%%
+%% %CopyrightEnd%
+%%
+
+-module(pubkey_ocsp).
+
+-include("public_key.hrl").
+
+-export([otp_cert/1, get_ocsp_responder_id/1, get_nonce/1, get_nonce_extn/1,
+         decode_ocsp_response/1, verify_ocsp_response/3,
+         get_acceptable_response_types_extn/0]).
+
+%%--------------------------------------------------------------------
+-spec verify_ocsp_response(binary(), list(), undefined | binary()) ->
+    {ok, term()} | {error, term()}.
+%%
+%% Description: Verify the OCSP response to get the certificate status
+%%--------------------------------------------------------------------
+verify_ocsp_response(OCSPResponseDer, ResponderCerts, Nonce) ->
+    do_verify_ocsp_response(
+        decode_ocsp_response(OCSPResponseDer), ResponderCerts, Nonce
+    ).
+
+%%--------------------------------------------------------------------
+-spec do_verify_ocsp_response({ok, #'BasicOCSPResponse'{}} | {error, term()},
+                              list(), undefined | binary()) ->
+    {ok, term()} | {error, term()}.
+%%
+%% Description: Verify the OCSP response to get the certificate status
+%%--------------------------------------------------------------------
+do_verify_ocsp_response(
+    {ok, #'BasicOCSPResponse'{
+        tbsResponseData = ResponseData,
+        signatureAlgorithm = SignatureAlgo,
+        signature = Signature,
+        certs = Certs
+    }}, ResponderCerts, Nonce) ->
+
+    #'ResponseData'{
+        responderID = ResponderID
+    } = ResponseData,
+
+    case verify_ocsp_signature(
+        public_key:der_encode('ResponseData', ResponseData),
+        SignatureAlgo#'AlgorithmIdentifier'.algorithm,
+        Signature, Certs ++ ResponderCerts,
+        ResponderID) of
```

Comment:
The certs attribute of the BasicOCSPResponse is meant to help us to verify the signature of the response:
```
   BasicOCSPResponse       ::= SEQUENCE {
      tbsResponseData      ResponseData,
      signatureAlgorithm   AlgorithmIdentifier,
      signature            BIT STRING,
      certs            [0] EXPLICIT SEQUENCE OF Certificate OPTIONAL }

   The value for signature SHALL be computed on the hash of the DER
   encoding of ResponseData.  The responder MAY include certificates in
   the certs field of BasicOCSPResponse that help the OCSP client verify
   the responder's signature.  If no certificates are included, then
   certs SHOULD be absent.
```
The ResponderCerts on the other hand is a list of certs that the client trusts with the following definition:
```
   In the OCSPStatusRequest, the "ResponderIDs" provides a list of OCSP
   responders that the client trusts.  A zero-length "responder_id_list"
   sequence has the special meaning that the responders are implicitly
   known to the server, e.g., by prior arrangement.
```
So we have multiple scenarios here, depending on if certs is empty or not, ocsp_responder_certs configured or not, and also if we trust the certs that we got in the BasicOCSPReponse. I'm thinking that maybe we need an additional options just for that.

**Scenario 1: certs is empty, ocsp_responder_certs not defined**
Responders implicitly known to the client -> use cacerts?

**Scenario 2: certs is not empty, ocsp_responder_certs not defined**
Responders implicitly known to the client -> use cacerts then use certs if we trust it? new option?

**Scenario 2: ocsp_responder_certs is defined**
Responders implicitly known to the client -> use only ocsp_responder_certs if cert not found check cacerts, if cert still not found check certs if we trust it?